python/repositories.bzl

```diff
@@ -381,17 +430,67 @@ def python_register_toolchains(
             **kwargs
         )
         if register_toolchains:
-            native.register_toolchains("@{name}_toolchains//:{platform}_toolchain".format(
-                name = name,
+            native.register_toolchains("@{toolchain_repo_name}//:{platform}_toolchain".format(
+                toolchain_repo_name = toolchain_repo_name,
                 platform = platform,
             ))
 
-    resolved_interpreter_os_alias(
-        name = name,
+    toolchains_repo(
+        name = toolchain_repo_name,
+        python_version = python_version,
+        set_python_version_constraint = set_python_version_constraint,
         user_repository_name = name,
     )
 
-    toolchains_repo(
-        name = "{name}_toolchains".format(name = name),
+    toolchain_aliases(
+        name = name,
+        python_version = python_version,
         user_repository_name = name,
     )
+
+def python_register_multi_toolchains(
+        name,
+        python_versions,
+        default_version = None,
+        **kwargs):
+    """Convenience macro for registering multiple Python toolchains.
+
+    Args:
+        name: base name for each name in python_register_toolchains call.
+        python_versions: the Python version.
+        default_version: the default Python version. If not set, the first version in
+            python_versions is used.
+        **kwargs: passed to each python_register_toolchains call.
+    """
+    if len(python_versions) == 0:
+        fail("python_versions must not be empty")
+
+    if not default_version:
+        default_version = python_versions.pop(0)
+    for python_version in python_versions:
+        if python_version == default_version:
+            # We register the default version lastly so that it's not picked first when --platforms
+            # is set with a constraint during toolchain resolution. This is due to the fact that
+            # Bazel will match the unconstrained toolchain if we register it before the constrained
+            # ones.
```

Comment:
Can you explain this a bit more? I don't fully understand the problem.

Each of the registered toolchains should have e.g. compatible_with=[X] such that they're mutually exclusive. When a rule does toolchain lookup, the rule should already have a python version set so that when Bazel matches its config to the available toolchains, it can only match one.

It sounds like the either the "mutually exclusive" part or the "rule's setting is set" part isn't true, and multiple toolchains are matching?

Reply:
Hm. Ok. Why is there a toolchain without a version constraint? Do we not know the python version when the toolchain is registered?